### DISCUSSION TITLE
Pluto Speed on Start-Up

### DISCUSSION THREAD

**Discussion by NuclearPowerNerd** [fonsp/Pluto.jl#851] [Questions about using Pluto]
Hello,

I am really enjoying learning Julia using Pluto.

One thing I have noticed though - when I first start up Pluto, even in a relatively small notebook, it takes several minutes to run all the cells and be ready for writing. Is that how it is supposed to work?

For reference, I am using Windows Subsystem for Linux 1 (WSL1) and starting pluto via the Windows Terminal while in WSL1. I run it out of a directory in which I've set up a Julia environment. Here is what the first few cells of the notebook look like. 

Am I doing something wrong? Or maybe its a WSL1 quirk? 

```
# Load packages
begin
        using Pkg
        Pkg.activate("Project.toml")
        using Plots, PlutoUI, OffsetArrays, LaTeXStrings
        gr() # load plotting backend
end
```

**Comment by BrandonConder** [comment]
How does it compare to running the same code in the REPL?

In my experience time to first plot is disappointingly bad, but not minutes.


Also just an FYI I've seen some blog posts suggesting improvements to precompilation are a big priority in Julia 1.6.0, so you may want to try the latest beta release.

**Reply by NuclearPowerNerd** [reply]
I've not tried just the REPL. Will try later tonight, see if it makes a difference. I'll also try a Pluto NB without any plots - maybe that is issue.

**Comment by dustyirwin** [comment]
This may not be a solution for your use case, but I've found that if you trace your notebook and them compile the image file (using PackageCompiler) the boot-up time is drastically decreased. In my case, some notebooks went from taking 1-2 minutes to just a few seconds.  

The basic compile process:

1. Start Julia/Pluto in trace mode: `julia --project --trace-compile="pluto_trace.jl" scripts/start_pluto.jl`
2. After you've run all the notebooks/cells you care about, compile your .so image file from the trace file:
`using PackagCompiler`
`create_sysimage(project_symbols,
    precompile_statements_file="pluto_trace.jl",
    sysimage_path="sys-pluto.so"
    )` where project_symbols is an array of the library names that you want to compile. eg project_symbols = [ :Plots, :Flux, ...].
3. Restart Julia/Pluto with your .so file: `julia -J"sys-pluto.so" scripts/start_pluto.jl`

The notebook should now load very fast. One caveat is that I've found these .so files to be fairly fragile. If you update your packages after compiling and then load the .so file, expect a bunch error messages and maybe a weird crash or two with some libraries (not sure why, maybe be an issue with my specific application). Good luck!
